driver-sync/src/test/functional/com/mongodb/client/AbstractTransactionsTest.java

```diff
@@ -455,4 +486,54 @@ private boolean canRunTests() {
             return false;
         }
     }
+
+    private class TargetedFailPoint {
+        private final BsonDocument failPointDocument;
+        private final MongoDatabase adminDB;
+
+        TargetedFailPoint(final BsonDocument operation) {
+            final BsonDocument arguments = operation.getDocument("arguments", new BsonDocument());
+            final ClientSession clientSession = sessionsMap.get(arguments.getString("session").getValue());
+
+            if (clientSession.getPinnedMongosAddress() != null) {
+                MongoClientSettings.Builder builder = MongoClientSettings.builder()
+                        .applyToClusterSettings(new Block<ClusterSettings.Builder>() {
+                            @Override
+                            public void apply(final ClusterSettings.Builder builder) {
+                                builder.hosts(singletonList(clientSession.getPinnedMongosAddress()));
+                            }
+                        });
+                MongoClient mongoClient = MongoClients.create(MongoClientSettings.builder(builder.build())
```

Comment:
Not necessary to create a builder, build it, create another builder, build it...  You can do it all with a single builder by chaining the apply methods to one another

Reply:
I combined code in the Java Driver documentation with code in the test runner, which was bad. I have restructured the code as suggested.